arangod/Aql/QueryRegistry.cpp

```diff
@@ -182,146 +188,201 @@ void* QueryRegistry::openEngine(EngineId id, EngineType type) {
-  WRITE_LOCKER(writeLocker, _lock);
 
-  auto it = _engines.find(engineId);
-  if (it == _engines.end()) {
-    LOG_TOPIC("97351", DEBUG, arangodb::Logger::AQL)
-        << "Found no engine with id " << engineId;
-    return;
-  }
+  // finishing the query will call back into the registry to destroy the query,
+  // so this must be done outside the lock
+  QueryInfo* queryInfoToFinish = nullptr;
 
-  EngineInfo& ei = it->second;
-  if (!ei._isOpen) {
-    LOG_TOPIC("45b8e", DEBUG, arangodb::Logger::AQL)
-        << "engine id " << engineId << " was not open.";
-    THROW_ARANGO_EXCEPTION_MESSAGE(
-        TRI_ERROR_INTERNAL, "engine with given vocbase and id is not open");
-  }
+  {
+    WRITE_LOCKER(writeLocker, _lock);
 
-  ei._isOpen = false;
+    auto it = _engines.find(engineId);
+    if (it == _engines.end()) {
+      LOG_TOPIC("97351", DEBUG, arangodb::Logger::AQL)
+          << "Found no engine with id " << engineId;
+      return;
+    }
 
-  if (ei._queryInfo) {
-    TRI_ASSERT(ei._queryInfo->_numOpen > 0);
-    ei._queryInfo->_numOpen--;
-    if (!ei._queryInfo->_query->killed() && ei._queryInfo->_expires != 0) {
-      ei._queryInfo->_expires = TRI_microtime() + ei._queryInfo->_timeToLive;
+    EngineInfo& ei = it->second;
+    if (!ei._isOpen) {
+      LOG_TOPIC("45b8e", DEBUG, arangodb::Logger::AQL)
+          << "engine id " << engineId << " was not open.";
+      THROW_ARANGO_EXCEPTION_MESSAGE(
+          TRI_ERROR_INTERNAL, "engine with given vocbase and id is not open");
+    }
+
+    ei._isOpen = false;
+
+    if (ei._queryInfo) {
+      TRI_ASSERT(ei._queryInfo->_numOpen > 0);
+      ei._queryInfo->_numOpen--;
+      if (ei._queryInfo->_numOpen == 0 && ei._queryInfo->_finished) {
+        // we were the last thread to close the engine, but the query has
+        // already been marked as finished, so we are responsible to resolve the
+        // future in order to actually finish the query
+        queryInfoToFinish = ei._queryInfo;
+      }
+      if (!ei._queryInfo->_query->killed() && ei._queryInfo->_expires != 0) {
+        ei._queryInfo->_expires = TRI_microtime() + ei._queryInfo->_timeToLive;
+      }
+      LOG_TOPIC("5ecdc", TRACE, arangodb::Logger::AQL)
+          << "closing engine " << engineId
+          << ", query id: " << ei._queryInfo->_query->id()
+          << ", numOpen: " << ei._queryInfo->_numOpen;
+    } else {
+      LOG_TOPIC("ae981", TRACE, arangodb::Logger::AQL)
+          << "closing engine " << engineId << ", no query";
     }
-    LOG_TOPIC("5ecdc", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId
-        << ", query id: " << ei._queryInfo->_query->id()
-        << ", numOpen: " << ei._queryInfo->_numOpen;
-  } else {
-    LOG_TOPIC("ae981", TRACE, arangodb::Logger::AQL)
-        << "closing engine " << engineId << ", no query";
+  }
+
+  if (queryInfoToFinish) {
+    queryInfoToFinish->_promise.setValue(queryInfoToFinish->_query);
```

Comment:
Is it safe to use the `_queryInfo` data here outside of the lock.
Can there be a race with the garbage collection of the QueryRegistry (haven't advanced this far down in this PR, but just wondering now when reading the changes from top to bottom).